File: 5-Problem-A-Day/298. Binary Tree Longest Consecutive Sequence.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode left; TreeNode
 * right; TreeNode(int x) { val = x; } }
 */
class Solution {

    int ans = 0;

    public int longestConsecutive(TreeNode root) {
        if (root == null)
            return 0;
        extend(root.left, root.val, 1);
        extend(root.right, root.val, 1);
        return ans;
    }

    private void extend(TreeNode root, int last, int length) {
        ans = Math.max(length, ans);
        if (root == null)
            return;
        if (root.val == last + 1)
            length += 1;
        else
            length = 1;
        extend(root.left, root.val, length);
        extend(root.right, root.val, length);
    }
}

File: 2021/298. Binary Tree Longest Consecutive Sequence.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxLen = 0;
    int longestConsecutive(TreeNode* root) {
        longest(root, INT_MIN, 0);
        return maxLen;
    }
    
    void longest(TreeNode* root, int parent, int len) { 
        maxLen = max(len, maxLen);
        if (root == NULL) return;        
        if (root->val != parent + 1) len = 1;
        else len += 1;
        longest(root->left, root->val, len);
        longest(root->right, root->val, len);
    }
};